lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -515,6 +515,54 @@ void TypeLoweringVisitor::visitDecl(MemOp op) {
   opsToRemove.push_back(op);
 }
 
+/// Copy annotations from \p annotations to \p loweredAttrs, except annotations
+/// with "target" key, that do not match the field suffix.
+static void filterAnnotations(ArrayAttr annotations,
+                              SmallVector<Attribute> &loweredAttrs,
+                              const FlatBundleFieldEntry &field,
+                              MLIRContext *context) {
+
+  for (auto opAttr : annotations) {
+    auto di = opAttr.dyn_cast<DictionaryAttr>();
+    if (!di) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+    auto targetAttr = di.get("target");
+    if (!targetAttr) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+
+    ArrayAttr subFieldTarget = targetAttr.cast<ArrayAttr>();
+    SmallString<16> targetStr;
+    for (auto fName : subFieldTarget) {
+      std::string fNameStr = fName.cast<StringAttr>().getValue().str();
+      // The fNameStr will begin with either '[' or '.', replace it with an
+      // '_' to construct the suffix.
+      fNameStr[0] = '_';
```

Comment:
This will annoyingly alias for the following circuit because the Scala FIRRTL Compiler allows for numeric fields:

```scala
circuit Baz: %[[{"a":null,"target":"~Baz|Baz>a[0]"}]]
  module Baz:
    wire a: {0: UInt<1>}
```

Note: this is a totally [reasonable assumption to make](https://github.com/chipsalliance/chisel3/pull/517#issuecomment-281817410) and is how it used to work before 2017.

This is very tedious to handle here because it means that `flattenType` needs to return _more_ information than it currently does. 

As I'd like to get this in, how about we do the following:
1. File a tracking issue about this.
2. We handle this with Target verification that runs as part of normal op verifiers (which asserts that all annotations point to valid things in that declaration).

I.e., I don't think this matters if we've pre-verified that the annotations are correct.

Reply:
#1013 was opened to track this.